llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp

```diff
@@ -634,59 +637,103 @@ void StatCollector::printToOuts(ArrayRef<CASID> TopLevels) {
-                          "", "Children", "", "Data (B)", "", "Cost (B)", "");
-  outs() << llvm::formatv(HeaderFormat.begin(), "====", "=====", "",
-                          "=======", "", "========", "", "========", "",
-                          "========", "");
-  auto printInfo = [Format, NumHashBytes, &Totals = this->Totals](
-                       StringRef Kind, ObjectKindInfo Info) {
-    if (!Info.Count)
-      return;
-    auto getPercent = [](double N, double D) { return D ? N / D : 0.0; };
-    size_t Size = Info.getTotalSize(NumHashBytes);
-    outs() << llvm::formatv(
-        Format.begin(), Kind, Info.Count, getPercent(Info.Count, Totals.Count),
-        Info.NumParents, getPercent(Info.NumParents, Totals.NumParents),
-        Info.NumChildren, getPercent(Info.NumChildren, Totals.NumChildren),
-        Info.DataSize, getPercent(Info.DataSize, Totals.DataSize), Size,
-        getPercent(Size, Totals.getTotalSize(NumHashBytes)));
-  };
-  for (StringRef Kind : Kinds)
-    printInfo(Kind, Stats.lookup(Kind));
-  printInfo("TOTAL", Totals);
-
-  // Other stats.
-  bool HasPrinted = false;
-  auto printIfNotZero = [&HasPrinted](StringRef Name, size_t Num) {
-    if (!Num)
-      return;
-    if (!HasPrinted)
-      outs() << "\n";
-    outs() << llvm::formatv("{0,-22} {1,+10}\n", Name, Num);
-    HasPrinted = true;
-  };
+  if (!ComputeStatsCSV.empty()) {
+    std::error_code EC;
+    raw_fd_ostream S(StringRef(ComputeStatsCSV), EC, sys::fs::OF_None);
+    S << "Kind,"
+      << "Count,"
+      << "Parents,"
+      << "Children,"
+      << "Data (B),"
+      << "Cost (B)"
+      << "\n";
+    auto printInfo = [&](StringRef Kind, ObjectKindInfo Info) {
+      if (!Info.Count)
+        return;
+      size_t Size = Info.getTotalSize(NumHashBytes);
+      S << Kind << ", " << Info.Count << ", " << Info.NumParents << ", "
+        << Info.NumChildren << ", " << Info.DataSize << ", " << Size << "\n";
+    };
+    for (StringRef Kind : Kinds)
+      printInfo(Kind, Stats.lookup(Kind));
+    printInfo("TOTAL", Totals);
+    auto printIfNotZero = [&](StringRef Name, size_t Num) {
+      if (!Num)
+        return;
+      S << Name << ", " << Num << "\n";
+    };
 
-  printIfNotZero("num-generated-names", GeneratedNames.size());
-  printIfNotZero("num-section-names", SectionNames.size());
-  printIfNotZero("num-symbol-names",
-                 SymbolNames.size() + UndefinedSymbols.size());
-  printIfNotZero("num-undefined-symbols", UndefinedSymbols.size());
-  printIfNotZero("num-anonymous-symbols", NumAnonymousSymbols);
-  printIfNotZero("num-template-symbols", NumTemplateSymbols);
-  printIfNotZero("num-template-targets", NumTemplateTargets);
-  printIfNotZero("num-zero-fill-blocks", NumZeroFillBlocks);
-  printIfNotZero("num-1-target-blocks", Num2TargetBlocks);
-  printIfNotZero("num-2-target-blocks", Num1TargetBlocks);
+    printIfNotZero("num-generated-names", GeneratedNames.size());
```

Comment:
Code is still quite duplicated here. I wonder if you can just print a `llvm::formatv` CSV output only since that is quite readable too. Just print that to stdout by default or `-` and write to a file when a path is given.

Reply:
I saw duncan's comments after I pushed this change, the code is completely different now